Buzz_gazebo/src/BuzzUtility.cpp

```diff
@@ -0,0 +1,246 @@
+void buzzScriptDestroy() {
+    if (VM->state != BUZZVM_STATE_READY)
+    {
+        ROS_ERROR("%s: execution terminated abnormally: %s", boFilename, buzzErrorInfo());
+        buzzvm_dump(VM);
+    }
+    buzzvm_function_call(VM, "destroy", 0);
+    buzzvm_destroy(&VM);
+    free(boFilename);
+    buzzdebug_destroy(&dbgInfo);
+    ROS_INFO("Script execution stopped.");
+}
+
+/*************************************************************************************************/
+int buzzScriptDone() {
+    return VM->state != BUZZVM_STATE_READY;
+}
+
+
+/*************************************************************************************************/
+int getRobotID() {
+    return robotID;
+}
+
+/*************************************************************************************************/
+int getSwarmSize() {
+    return (int)buzzdict_size(VM->swarmmembers) + 1;
+}
+
+/*************************************************************************************************/
+buzzvm_t getVM() {
+    return VM;
+}
+
+/*************************************************************************************************/
+std::string getVMState() {
+    std::string state = "Unknown";
+    if (VM && VM->strings)
+    {
+    buzzvm_pushs(VM, buzzvm_string_register(VM, "BVMSTATE", 1));
+    buzzvm_gload(VM);
+    buzzobj_t obj = buzzvm_stack_at(VM, 1);
+    if (obj->o.type == BUZZTYPE_STRING)
+        state = std::string(obj->s.value.str);
+    else
+        state = "Not Available";
+    buzzvm_pop(VM);
+    }
+    return state;
+}
+
+/*************************************************************************************************/
+const char* buzzErrorInfo() {
+  buzzdebug_entry_t dbg = *buzzdebug_info_get_fromoffset(dbgInfo, &VM->pc);
+  char* msg;
+  if (dbg != NULL)
+  {
+    asprintf(&msg, "%s: execution terminated abnormally at %s:%" PRIu64 ":%" PRIu64 " : %s\n\n", boFilename, dbg->fname,
+             dbg->line, dbg->col, VM->errormsg);
+  }
+  else
+  {
+    asprintf(&msg, "%s: execution terminated abnormally at bytecode offset %d: %s\n\n", boFilename, VM->pc, VM->errormsg);
+  }
+  return msg;
+}
+
+/*************************************************************************************************/
+std::string compileBuzzScript(std::string p_bzzFilename){
+    std::stringstream bzzfile_in_compile;
+    std::string path = p_bzzFilename.substr(0, p_bzzFilename.find_last_of("\\/")) + "/";
+    std::string name = p_bzzFilename.substr(p_bzzFilename.find_last_of("/\\") + 1);
+    name = name.substr(0, name.find_last_of("."));
+    bzzfile_in_compile << "bzzc -I " << path << "include/";
+    bzzfile_in_compile << " -b " << path << name << ".bo";
+    bzzfile_in_compile << " -d " << path << name << ".bdb ";
+    bzzfile_in_compile << p_bzzFilename;
+
+    ROS_WARN("Launching buzz compilation: %s", bzzfile_in_compile.str().c_str());
+
+    system(bzzfile_in_compile.str().c_str());
```

Comment:
Can we get the return code so we know the compilation worked?

Reply:
Function now returns "error" if compilation failed so that the bridge may kill the node with its known name